### DISCUSSION TITLE
Quantile Binning Refactor

### DISCUSSION THREAD

**Discussion by Schobs** [pykale/pykale#392] [General]
Discussion for TODO items on the QBin Refactor.
Currently @ PR [here](https://github.com/pykale/pykale/pull/281) .

TODO:

- [ ] boxplot display settings as params in interpret/uncertainty_quantiles functions.
- [ ] Circular import issue in evaluate_correlations(). Line 98 of kale/evaluate/similarity_metrics.py
- [ ] Refactor of evaluate/uncertainty_metrics for repeated code blocks.
- [ ] Refactor of interpet/uncertainty_quantiles for repeated code. Possibly make Boxplot class.
- [ ] Consider replacing custom Jaccard with [sklearn learn jaccard](https://scikit-learn.org/stable/modules/generated/sklearn.metrics.jaccard_score.html).
- [ ] There are repeated for loops and fragments, could you implement them as functions/classes to reuse?
- [ ] `warnings.filterwarnings("error")` at `examples/landmark_uncertainty/main.py`
     - [ ] Do you have to put this between imports?
     - [ ] Also, is ignoring errors a good practice?
- [ ] Improve Code Coverage for:
     - [ ]  similarity_metrics.py -  saving figs in evaluate_correlations() crashes on Github Actions
     - [ ] uncertainty_metrics.py - saving figs in fit_line_with_ci crashes on Github Actions

**Comment by kungfudaibi** [comment]
* [x]  boxplot display settings as params in interpret/uncertainty_quantiles functions.
Created a `BoxplotStyleConfig` class to extract all hard-coded values from the code. You can adjust the style by instantiating this class.

* [x]  Circular import issue in evaluate_correlations(). Line 98 of kale/evaluate/similarity_metrics.py
moved the `fit_line_with_ci` function out of `kale/interpret/uncertainty_quantiles.py` into a separate standalone file, it looks only `kale/evaluate/similarity_metrics.py` use this function and i  put it in `kale/interpret/correlation_analysis.py`(maybe splitting uncertainty_metrics.py would make the code more modular？)

* [x]  Refactor of evaluate/uncertainty_metrics for repeated code blocks.
Created several classes to improve readability and reduce repetitive code blocks.Because some of the loop blocks have many subtle differences, most of the base class methods had to be overridden. As a result, the amount of duplicated code reduced isn't substantial yet, and the helper functions haven’t been refactored either.

* [x]  Refactor of interpet/uncertainty_quantiles for repeated code. Possibly make Boxplot class.
Created a `Box_plot` class for Plot boxplots: generic_box_plot_loop, format_plot, box_plot_per_model, box_plot_comparing_q to reduce duplicate code

* [ ]  Consider replacing custom Jaccard with [sklearn learn jaccard](https://scikit-learn.org/stable/modules/generated/sklearn.metrics.jaccard_score.html).
* [x]  There are repeated for loops and fragments, could you implement them as functions/classes to reuse?
* [x]  `warnings.filterwarnings("error")` at `examples/landmark_uncertainty/main.py`
As we all know, programmers tend to ignore warnings, so I set `warnings.filterwarnings('error', category=RuntimeWarning)`

* [ ]  Improve Code Coverage for:
   
  - [ ]   similarity_metrics.py -  saving figs in evaluate_correlations() crashes on Github Actions
  - [ ]  uncertainty_metrics.py - saving figs in fit_line_with_ci crashes on Github Actions

current pr #485
